arctic/async/async_arctic.py

```diff
@@ -0,0 +1,187 @@
+
+
+def _arctic_task_exec(request):
+    request.start_time = time.time()
+    logging.debug("Executing asynchronous request for {}/{}".format(request.library, request.symbol))
+    result = None
+    try:
+        request.is_running = True
+        if request.mongo_retry:
+            result = mongo_retry(request.fun)(*request.args, **request.kwargs)
+        else:
+            result = request.fun(*request.args, **request.kwargs)
+    except Exception as e:
+        request.exception = e
+    finally:
+        request.data = result
+        request.end_time = time.time()
+        request.is_running = False
+    return result
+
+
+class AsyncArctic(LazySingletonWorkersPool):
+    _instance = None
+    _SINGLETON_LOCK = RLock()
+    _POOL_LOCK = RLock()
+
+    def __init__(self, pool_size):
+        # Only allow creation via get_instance
+        if not type(self)._SINGLETON_LOCK._is_owned():
+            raise AsyncArcticException("AsyncArctic is a singleton, can't create a new instance")
+
+        # Enforce the singleton pattern
+        with type(self)._SINGLETON_LOCK:
+            super(AsyncArctic, self).__init__(pool_size)
+            self.requests_per_library = defaultdict(lambda: defaultdict(lambda: defaultdict(list)))
+            self.requests_by_id = dict()
+
+    def __reduce__(self):
+        return "ASYNC_ARCTIC"
+
+    def _get_modifiers(self, library_name, symbol=None):
+        return self.requests_per_library[library_name][symbol][AsyncRequestType.MODIFIER]
+
+    def _get_accessors(self, library_name, symbol=None):
+        return self.requests_per_library[library_name][symbol][AsyncRequestType.ACCESSOR]
+
+    @staticmethod
+    def _verify_request(store, is_modifier, **kwargs):
+        library_name = None if store is None else store._arctic_lib.get_name()
+        symbol = kwargs.get('symbol')
+        kind = AsyncRequestType.MODIFIER if is_modifier else AsyncRequestType.ACCESSOR
+        callback = kwargs.get('async_callback')
+        mongo_retry = bool(kwargs.get('mongo_retry'))
+        return library_name, symbol, kind, callback, mongo_retry
+
+    def _add_request(self, request):
+        self.requests_per_library[request.library][request.symbol][request.kind].append(request)
+        self.requests_by_id[request.id] = request
+
+    def _remove_request(self, request):
+        self.requests_per_library[request.library][request.symbol][request.kind].remove(request)
+        if request.id in self.requests_by_id:
+            del self.requests_by_id[request.id]
+
+    def submit_arctic_request(self, store, fun, is_modifier, *args, **kwargs):
+        lib_name, symbol, kind, callback, mongo_retry = AsyncArctic._verify_request(store, is_modifier, **kwargs)
+
+        for k in ('async_callback', 'mongo_retry'):
+            kwargs.pop(k, None)
+
+        with type(self)._POOL_LOCK:  # class level lock, since it is a Singleton
+            if lib_name:
+                if self._get_modifiers(lib_name, symbol):
```

Comment:
It seems awkward that in async code clients need to deal with this themselves.  Shouldn't the library manage this for callers?

What can a client processing a request do when receiving this error partially through a larger operation?  Seems like all that caller can do is make a *synchronous* blocking retry to continue.

Reply:
I would suggest deferring this part until we have a solid plan.  You can add locking above, but also it doesn't seem like it will stop a concurrent synchronous caller.  Nor a mutator in a different process/host.

Reply:
Valid points, as we discussed, I will convert this to an execution deferral approach.

So we have a queue of jobs awaiting to be run which will get delayed as long as a conflicting task is running currently in the pool, to avoid concurrent modification errors like the ones described here: #655 

Note, we guaranty safety for concurrent read/write for the same symbol from multiple threads in the same process here, not across processes.